core/src/connection/pool.rs

```diff
@@ -0,0 +1,907 @@
+    ///
+    /// Returns the assigned connection ID on success. An error is returned
+    /// if the configured maximum number of established connections for the
+    /// connected peer has been reached.
+    pub fn add<TMuxer>(&mut self, c: Connection<TMuxer, THandler::Handler>, i: Connected<TConnInfo>)
+        -> Result<ConnectionId, ConnectionLimit>
+    where
+        THandler: IntoConnectionHandler<TConnInfo> + Send + 'static,
+        THandler::Handler: ConnectionHandler<
+            Substream = connection::Substream<TMuxer>,
+            InEvent = TInEvent,
+            OutEvent = TOutEvent,
+            Error = THandlerErr
+        > + Send + 'static,
+        <THandler::Handler as ConnectionHandler>::OutboundOpenInfo: Send + 'static,
+        TTransErr: error::Error + Send + 'static,
+        THandlerErr: error::Error + Send + 'static,
+        TInEvent: Send + 'static,
+        TOutEvent: Send + 'static,
+        TMuxer: StreamMuxer + Send + Sync + 'static,
+        TMuxer::OutboundSubstream: Send + 'static,
+        TConnInfo: Clone + Send + 'static,
+        TPeerId: Clone,
+    {
+        if let Some(limit) = self.limits.max_established_per_peer {
+            let current = self.num_peer_established(i.peer_id());
+            if limit >= current {
+                return Err(ConnectionLimit { limit, current })
+            }
+        }
+        let id = self.manager.add(c, i.clone());
+        self.established.entry(i.peer_id().clone()).or_default().insert(id, i.endpoint);
+        Ok(id)
+    }
+
+    /// Gets an entry representing a connection in the pool.
+    ///
+    /// Returns `None` if the pool has no connection with the given ID.
+    pub fn get(&mut self, id: ConnectionId)
+        -> Option<PoolConnection<'_, TInEvent, TConnInfo, TPeerId>>
+    {
+        match self.manager.entry(id) {
+            Some(manager::Entry::Established(entry)) =>
+                Some(PoolConnection::Established(EstablishedConnection {
+                    entry,
+                    established: &mut self.established,
+                })),
+            Some(manager::Entry::Pending(entry)) =>
+                Some(PoolConnection::Pending(PendingConnection {
+                    entry,
+                    pending: &mut self.pending,
+                })),
+            None => None
+        }
+    }
+
+    /// Gets an established connection from the pool.
+    ///
+    /// If a connection ID is given and no established connection with that
+    /// ID is known to the pool, `None` is returned.
+    ///
+    /// If no connection ID is given and the pool contains at least one
+    /// established connection to the given peer, an unspecified selection
+    /// is made among these connections.
+    ///
+    /// If no connection ID is given and no established connection exists
+    /// for the given peer, `None` is returned.
+    pub fn get_established(&mut self, peer: &TPeerId, id: Option<ConnectionId>)
```

Comment:
That is an extremely weird API. I'd prefer if this function didn't exist and the user has to explicitly select which connection is wanted.

Reply:
Ok, done.